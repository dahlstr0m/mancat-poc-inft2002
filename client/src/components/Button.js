// @flow

import * as React from 'react';
import { Component } from 'react-simplified';

/**
 * Renders a success button using Bootstrap styles.
 */
class ButtonSuccess extends Component<{
  onClick: () => mixed,
  small?: boolean,
  children?: React.Node,
}> {
  render() {
    return (
      <button
        type="button"
        className={'btn btn-success' + (this.props.small ? ' btn-sm py-0' : '')}
        onClick={this.props.onClick}
      >
        {this.props.children}
      </button>
    );
  }
}

/**
 * Renders a danger button using Bootstrap styles.
 */
class ButtonDanger extends Component<{
  onClick: () => mixed,
  small?: boolean,
  children?: React.Node,
}> {
  render() {
    return (
      <button
        type="button"
        className={'btn btn-danger' + (this.props.small ? ' btn-sm py-0' : '')}
        onClick={this.props.onClick}
      >
        {this.props.children}
      </button>
    );
  }
}

/**
 * Renders a light button using Bootstrap styles.
 */
class ButtonLight extends Component<{
  onClick: () => mixed,
  small?: boolean,
  children?: React.Node,
}> {
  render() {
    return (
      <button
        type="button"
        className={'btn btn-light' + (this.props.small ? ' btn-sm py-0' : '')}
        onClick={this.props.onClick}
      >
        {this.props.children}
      </button>
    );
  }
}

/**
 * Renders a button using Bootstrap styles.
 */
export default class Button {
  static Success = ButtonSuccess;
  static Danger = ButtonDanger;
  static Light = ButtonLight;
}
