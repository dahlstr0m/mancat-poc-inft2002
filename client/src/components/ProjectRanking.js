// @flow

import * as React from 'react';
import { Component } from 'react-simplified';
import projectService, { type Project } from '../services/project-service';
import posterService, { type Poster } from '../services/poster-service';
import { Card } from './Card';
import Form from './Form';
import { Alert } from './Widgets';
import Button from './Button';

/**
 * Renders ranking list
 */

export default class ProjectRanking extends Component {
  projects: Project[] = [];
  posters: Poster[] = [];
  rankIndex: Project[] = [];

  render() {
    return (
      <>
        <Card title="Change projects display-order">
          <Button.Success onClick={() => projectService.updateRanking(this.rankIndex)}>
            Save ranking
          </Button.Success>
          <Card>
            <table className="table">
              <thead>
                <tr>
                  <th scope="col">Projects vieworder</th>
                  <th scope="col">Project</th>
                  <th scope="col">Status</th>
                  <th scope="col">Move</th>
                </tr>
              </thead>
              <tbody>
                {this.rankIndex.map((
                  project // Mapping to display all projects with Ranking.
                ) => (
                  <tr key={project.projectId}>
                    <th scope="row">{'#' + project.ranking}</th>

                    <td>
                      <span>
                        <img
                          src={this.getPosterUrl(project)}
                          alt="No thumb"
                          width="53"
                          height="75"
                        />
                      </span>
                      <span>{' ' + project.title}</span>
                    </td>
                    <td>{project.active === true ? 'Active' : 'Disabled'}</td>
                    <td>
                      <Button.Light small onClick={() => this.rankUp(project)}>
                        Up
                      </Button.Light>{' '}
                      |{' '}
                      <Button.Light small onClick={() => this.rankDown(project)}>
                        Down
                      </Button.Light>
                    </td>
                  </tr>
                ))}
              </tbody>
            </table>
          </Card>
        </Card>
      </>
    );
  }

  rankUp(project: Project) {
    if (project.ranking <= 1) {
      Alert.danger('Can not rank project any higher.');
      return;
    }
    const projectIndex = project.ranking - 1;
    this.rankIndex[projectIndex].ranking -= 1;
    this.rankIndex[projectIndex - 1].ranking += 1;

    this.rankIndex = this.rankIndex.sort(
      (projectA, projectB) => projectA.ranking - projectB.ranking
    );
  }

  rankDown(project: Project) {
    if (project.ranking >= this.rankIndex.length) {
      Alert.danger('Can not rank project any lower.');
      return;
    }
    const projectIndex = project.ranking - 1;
    this.rankIndex[projectIndex].ranking += 1;
    this.rankIndex[projectIndex + 1].ranking -= 1;

    this.rankIndex = this.rankIndex.sort(
      (projectA, projectB) => projectA.ranking - projectB.ranking
    );
  }

  getPosterUrl(project: Project) {
    const poster = this.posters.find((poster) => poster.projectId === project.projectId);
    return poster ? poster.thumbnailUrl : '';
  }

  sortProjectRanking() {
    this.rankIndex = [];
    this.projects.forEach((project, index) => {
      const newProject = project;
      newProject.ranking = index + 1;
      this.rankIndex.push(newProject);
    });
  }

  mounted() {
    projectService
      .getProjects()
      .then((projects) => (this.projects = projects))
      .then(() =>
        this.projects.forEach((project) => {
          if (project.ranking === 0) {
            project.ranking = project.projectId;
          }
        })
      )
      .then(() => this.projects.sort((projectA, projectB) => projectA.ranking - projectB.ranking)) // Sorting to display by asc ranking
      .then(() => this.sortProjectRanking())
      .catch((error: Error) => Alert.danger('Error getting projects: ' + error.message));

    posterService
      .getPosters()
      .then((posters) => (this.posters = posters))
      .catch((error: Error) => Alert.danger('Error getting posters: ' + error.message));
  }
}
