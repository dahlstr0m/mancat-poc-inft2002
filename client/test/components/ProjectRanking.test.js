// @flow

import * as React from 'react';
import { shallow } from 'enzyme';

// Components
import ProjectRanking from '../../src/components/ProjectRanking';
import Button from '../../src/components/Button';

// Mocks
jest.mock('../../src/services/project-service');
jest.mock('../../src/services/poster-service');

// Tests
describe('ProjectRanking component tests', () => {
  test('ProjectRanking draws correctly', (done) => {
    const wrapper = shallow(<ProjectRanking />);

    // Wait for events to complete
    setTimeout(() => {
      expect(wrapper).toMatchSnapshot();

      done();
    });
  });

  test('ProjectRanking rankDown works as expected', (done) => {
    const wrapper = shallow(<ProjectRanking />);

    setTimeout(() => {
      expect(
        wrapper
          .find('img')
          .at(0)
          .equals(<img src="Poster 1 thumb url" alt="No thumb" width="53" height="75" />)
      ).toEqual(true);

      wrapper.find(Button.Light).at(1).simulate('click');
    });

    setTimeout(() => {
      console.log(wrapper.find('img').at(0).debug());
      expect(
        wrapper
          .find('img')
          .at(0)
          .equals(<img src="Poster 2 thumb url" alt="No thumb" width="53" height="75" />)
      ).toEqual(true);
      done();
    });
  });

  test('ProjectRanking rankUp works as expected', (done) => {
    const wrapper = shallow(<ProjectRanking />);

    setTimeout(() => {
      expect(
        wrapper
          .find('img')
          .at(0)
          .equals(<img src="Poster 1 thumb url" alt="No thumb" width="53" height="75" />)
      ).toEqual(true);

      wrapper.find(Button.Light).at(2).simulate('click');
    });

    setTimeout(() => {
      console.log(wrapper.find('img').at(0).debug());
      expect(
        wrapper
          .find('img')
          .at(0)
          .equals(<img src="Poster 2 thumb url" alt="No thumb" width="53" height="75" />)
      ).toEqual(true);
      done();
    });
  });
});
