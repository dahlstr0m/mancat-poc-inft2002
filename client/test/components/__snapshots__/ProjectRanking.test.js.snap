// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectRanking component tests ProjectRanking draws correctly 1`] = `
<Fragment>
  <Card
    title="Change projects display-order"
  >
    <ButtonSuccess
      onClick={[Function]}
    >
      Save ranking
    </ButtonSuccess>
    <Card>
      <table
        className="table"
      >
        <thead>
          <tr>
            <th
              scope="col"
            >
              Projects vieworder
            </th>
            <th
              scope="col"
            >
              Project
            </th>
            <th
              scope="col"
            >
              Status
            </th>
            <th
              scope="col"
            >
              Move
            </th>
          </tr>
        </thead>
        <tbody>
          <tr
            key="1"
          >
            <th
              scope="row"
            >
              #1
            </th>
            <td>
              <span>
                <img
                  alt="No thumb"
                  height="75"
                  src="Poster 1 thumb url"
                  width="53"
                />
              </span>
              <span>
                 Project 1
              </span>
            </td>
            <td>
              Active
            </td>
            <td>
              <ButtonLight
                onClick={[Function]}
                small={true}
              >
                Up
              </ButtonLight>
               
              |
               
              <ButtonLight
                onClick={[Function]}
                small={true}
              >
                Down
              </ButtonLight>
            </td>
          </tr>
          <tr
            key="2"
          >
            <th
              scope="row"
            >
              #2
            </th>
            <td>
              <span>
                <img
                  alt="No thumb"
                  height="75"
                  src="Poster 2 thumb url"
                  width="53"
                />
              </span>
              <span>
                 Project 2
              </span>
            </td>
            <td>
              Disabled
            </td>
            <td>
              <ButtonLight
                onClick={[Function]}
                small={true}
              >
                Up
              </ButtonLight>
               
              |
               
              <ButtonLight
                onClick={[Function]}
                small={true}
              >
                Down
              </ButtonLight>
            </td>
          </tr>
          <tr
            key="3"
          >
            <th
              scope="row"
            >
              #3
            </th>
            <td>
              <span>
                <img
                  alt="No thumb"
                  height="75"
                  src="Poster 3 thumb url"
                  width="53"
                />
              </span>
              <span>
                 Project 3
              </span>
            </td>
            <td>
              Active
            </td>
            <td>
              <ButtonLight
                onClick={[Function]}
                small={true}
              >
                Up
              </ButtonLight>
               
              |
               
              <ButtonLight
                onClick={[Function]}
                small={true}
              >
                Down
              </ButtonLight>
            </td>
          </tr>
        </tbody>
      </table>
    </Card>
  </Card>
</Fragment>
`;
